<a href="https://colab.research.google.com/github/Anuradha-Yadav31/Hello-World/blob/main/Anuradha_Yadav.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Importing Libraries:
Importing some of the libraries to be used in the module.


In [ ]:
import pandas as pd
import numpy as np


# Exploring the Data:
The data we will use is the iris dataset, a classical dataset in machine learning and statistics. It is included in scikit-learn in the datasets module. We can load it by the load_iris function.


In [ ]:
from sklearn.datasets import load_iris
iris_dataset = load_iris()

The iris object that is returned by load_iris is a Bunch object, which is very similar to a dictionary. It contains keys and values.

In [ ]:
print ("Keys of iris_dataset: \n{}".format(iris_dataset.keys()))

Keys of iris_dataset: 
dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names', 'filename'])


###The value of the key DESCR is a short description of the dataset.
Lets look at what it contains:

In [ ]:
print(iris_dataset['DESCR'][:193] + "\n...")

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, pre
...


As we can see it contains some information about the dataset itself, feel free to go through it to understand more about the dataset.

###Lets move on to the target_names.

The value of the key target_names is an array of strings, containing the species of flowers that we want to predict i.e. 'setosa', 'versicolour' and 'virginica'.

In [ ]:
print(" Target names: {}".format(iris_dataset['target_names']))

 Target names: ['setosa' 'versicolor' 'virginica']


###Next is the feature_names.
The value of feature names is the list of strings, giving the desription of each feature it includes the 'sepal length', 'sepal width', 'petal length' and tha 'petal width' all in centimeters.

In [ ]:
print("Feature names: \n{}".format(iris_dataset['feature_names']))

Feature names: 
['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


###Now lets see data.
The data that we are going to use is itself contained in the target and data fields. The data contains numeric measurements of sepal length, sepal width, petal lenght and petal width in a NumPy array.

In [ ]:
print("Type of data: {}".format(type(iris_dataset['data'])))

Type of data: <class 'numpy.ndarray'>


The rows in this data array corresponds to the flowers, while the column represent the four measurements that were taken for each flower: 

In [ ]:
print("Shape of data: {}".format(iris_dataset['data'].shape))

Shape of data: (150, 4)


The array contains measurements for 150 different flowers with 4 attributes each. Remember that the individual items are called samples/instances/data points in machine learning and their properties are called attributes/features. So we have 150 data points and 4 features. The shape of the data array is the number of samples multiplied by the number of features.

Here are the feature values for the first five samples:

In [ ]:
print("First five rows of data:\n {}".format(iris_dataset['data'][:5]))

First five rows of data:
 [[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]]


From this data, we can see that all of the first five flowers have petal width 0f 0.2 cm and that the longest sepal, at 5.1 cm. 

###Lets come to the last key i.e. the Target array.
Contains the species of each of the flowers that were measured, also as a NumPy array:

In [ ]:
print ("Type of Target: {}".format(type(iris_dataset['target'])))

Type of Target: <class 'numpy.ndarray'>


Target is a one-dimensional array, and on seeing the shape we can see that it contains one entry per flower:

In [ ]:
print("Shape of target: {}".format(iris_dataset['target'].shape ))

Shape of target: (150,)


On seeing the target key and exploring the values we can easily see that the species are encoded as integers from 0 to 2:

In [ ]:
print("Target: \n {}".format(iris_dataset['target']))

Target: 
 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


The meanings of the numbers are given by the iris['target_names'] array: 0 means setosa, 1 means versicolour and 2 means virginica.

#Measuring Success:Training and Testing Data:

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
           iris_dataset['data'], iris_dataset['target'], random_state=0)

The output of the train_test_split function is X_train, X_test, y_train and y_test, which are all NumPy arrays. X_train contains 75% of the rows of the dataset, and X_test contains the remaining 25%:

In [ ]:
print("X_train shape: {}".format(X_train.shape))
print("X_test shape: {}".format(X_test.shape))
print("y_train shape: {}".format(y_train.shape))
print("y_test shape: {}".format(y_test.shape))

X_train shape: (112, 4)
X_test shape: (38, 4)
y_train shape: (112,)
y_test shape: (38,)


#K- nearest neighbor machine learning model:

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=1)

The knn object encapsulates the algorithm that will be used to build the model from the training data, as well the algorithm to make predictions on new points.

We have trained the model on the training data.

In [ ]:
knn.fit(X_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=1, p=2,
                     weights='uniform')

#Making Predictions

We can now make predictions using this model on new model.
For Example- A flower with sepal length of 5 cm, sepal width of 2.9 cm, petal length of 1 cm and petal width of 0.2 cm.

In [ ]:
X_new = np.array([[5,2.9,1,0.2]])
print("X_new.shape: {}".format(X_new.shape))

X_new.shape: (1, 4)


To make a prediction, we call the predict method of the knn object:

In [ ]:
prediction = knn.predict(X_new)
print("Prediction: {}". format(prediction))
print("Predicted target name: {}".format(
    iris_dataset['target_names'][prediction]))

Prediction: [0]
Predicted target name: ['setosa']


We don't know the correct species of this sample, which is the whole point of building the model!

#Evaluating the Model:

In [ ]:
y_pred =knn.predict(X_test)
print("Test set predictions: \n {}".format(y_pred))


Test set predictions: 
 [2 1 0 2 0 2 0 1 1 1 2 1 1 1 1 0 1 1 0 0 2 1 0 0 2 0 0 1 1 0 2 1 0 2 2 1 0
 2]


In [ ]:
print("Test set score: {}".format(np.mean(y_pred== y_test)))

Test set score: 0.9736842105263158
